MITLibraries/cdn-auth-geo
# Adds SAML SP

Why are these changes being introduced:

* A SAML SP is method we use to interact with MIT Touchstone authentication

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/GDT-121

How does this address that need:

* Borrowing heavily from our previous application `bd_auth`, this implements a SAML SP via `python3-saml`
* a new make command `make app-bash` was created to streamline local development of this application while avoiding a problematic dependency
* To test these changes locally
  * should create an SP cert/key pair and extract them to single lines with no spaces and set in your local `.env` file (see README for the env values to set these to)
  * start the local IdP (see README)
  * start a local shell for the app with `make app-bash`
  * start the local dev server in the shell from the previous state with `make run-dev`
  * access file that should allow for redirects such as `http://localhost:5000/?cdn_resource=https://cdn.libraries.mit.edu/fakefile`. You should be prompted to authenticate (credentials in README) and then redirect to a "file not found" page for the CDN. You cannot access actually restricted files in the CDN when running locally like this as we pass JWT tokens as domain cookies which cannot be read. You _could_ set a valid domain cookie by modifying /etc/hosts and running your local dev environment on something like `totallyrealserver.libraries.mit.edu`. You'd need to pull the appropriate JWT secret from SSM.
  * See the tests for other URL patterns you may want to try (like leaving out the ?cdn_resource query param or setting it to a value that we won't redirect to for security reasons, etc)

What is missing:

* unit tests for SAML interactions. We should add unit tests for the auth.py module, but for now we need to manually test in our staging environment before moving code to production. As unit tests won't test against live Touchstone, this manual test will remain necessary even after we have automated tests in place.

Document any side effects to this change:

* Development is now expected to be in docker as libxmlsec1 dependency is complicated (at best) to get installed reliably locally (the upstream libraries are slowly working through the bugs and hope to have fixes in April 2024 which should allow us to move back to non-dockerized development if preferred)
* mypy linter was removed as we aren't using types in this application
